### ISSUE TITLE
Non-generic Fake and Dummy creation

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#482]
I just realized that `A.Fake` doesn't have a non-generic overload accepting a `Type` parameter, which would be very useful in some scenarios. I realize that it's not something most people would need on a regular basis, but there doesn't seem to be a good workaround. The only way I found was to call `A.Fake<T>()` via reflection, which is far from ideal.

Am I missing something?

Perhaps adding a non-generic overload to `A.Fake` wouldn't be the best option, because would pollute the API with something that is not usually useful, but it would be nice to have an easy workaround.

**Comment by blairconrad** [comment]
Hi, @thomaslevesque. Is there a particular use case you envision? Have you encountered a need for this, or is it just speculation? About a year ago, #288 was raised, which is nearly the same, but for `A.Dummy`. It was ultimately closed "wontfix" after we discussed it with the potential contributor and it turns out that even he considered it a niche feature.

**Comment by thomaslevesque** [comment]
Hi @blairconrad,

Yes, I have a specific use case in mind: http://stackoverflow.com/a/29905525/98713
As you can see I had to use `GetMethod` + `MakeGenericMethod` to achieve what I wanted.

I realize it's not a common use case, but it's not so far-fetched either, and currently there is no clean way to do it, as far as I can see. It's probably you don't really need it when you use FakeItEasy directly, but as soon as you build some tooling around it, it becomes almost indispensable.

**Comment by thomaslevesque** [comment]
Interestingly, it's kind of the opposite of #470, where I wanted to add a generic version of a method that accepted a `Type` parameter ;)
It's common practice to offer both generic and non-generic forms of an API that accepts a type, because depending on the use case, you might need one of the other.

**Comment by thomaslevesque** [comment]
The only issue that I see is that this non-generic overload couldn't accept an `Action<IFakeOptionsBuilder<T>>`, since there would be no `T`... (in my use case, it's an acceptable limitation)

**Comment by blairconrad** [comment]
Ah, yes. Auto-faking. I added some comments in the StackOverflow question.
Generally in these sorts of situations, I'd recommend AutoFixture/AutoFixture, especially with its AutoFakeItEasy extension.
It's _built_ for this sort of thing, and I'm sure @ploeh and contributors have added all kinds of nice features. There are even extensions for other containers, such as [the one for Autofac](http://docs.autofac.org/en/latest/integration/fakeiteasy.html).

Or, you might want to try `UnderTest` and `Fake`. Maybe they'll get you where you want to go, although they aren't as advanced.

**Comment by ploeh** [comment]
FWIW, here's a [pattern description of Auto-Mocking Containers](http://blog.ploeh.dk/2013/03/11/auto-mocking-container), with a general, library-independent description of the motivation, but a specific example using Castle Windsor and Moq.

Essentially, what you'll need is any Composition Engine with enough extensibility to hook into, and a Dynamic Mock library - also with a bit of extensibility.

For [AutoFixture](https://github.com/AutoFixture/AutoFixture), we've found it possible to create Auto-Mocking extensions for all major Dynamic Mock libraries for .NET:

- [AutoFixture.AutoMoq](http://nuget.org/packages/AutoFixture.AutoMoq)
- [AutoFixture.AutoRhinoMocks](http://nuget.org/packages/AutoFixture.AutoRhinoMocks)
- [AutoFixture.AutoFakeItEasy](http://nuget.org/packages/AutoFixture.AutoFakeItEasy)
- [AutoFixture.AutoNSubstitute](http://nuget.org/packages/AutoFixture.AutoNSubstitute)
- [AutoFixture.AutoFoq](http://www.nuget.org/packages/AutoFixture.AutoFoq)

The point here being that we have found all of these Dynamic Mock libraries to be extensible *enough* that it was possible to implement these packages.

As described in the pattern article, you can also base an Auto-Mocking Container on a DI Container. [My book](http://amzn.to/12p90MG) covers six DI Containers in detail, so you may find inspiration there.

**Comment by thomaslevesque** [comment]
Yes, AutoFixture seems to do exactly what I want. I was looking for a solution based on Unity, but actually it doesn't really matter.

**Comment by thomaslevesque** [comment]
But I still think a non-generic `A.Fake` would be useful to build tools like AutoFixture. If you're worried about polluting the API, you can always use the `[EditorBrowsable]` attribute to hide it from Intellisense.

**Comment by ploeh** [comment]
> If you're worried about polluting the API, you can always use the [EditorBrowsable] attribute to hide it from Intellisense.

FWIW, I've been there. That tends to generally not help. You can change your Visual Studio settings so that such members still show up, and IIRC; ReSharper doesn't respect that attribute at all.

**Comment by blairconrad** [comment]
Indeed, ReSharper doesn't. That drove me bonkers for about an hour one evening, as I was sure I was doing _something_ wrong the first time I tried to apply it to a member.

If one's building a tool such as AutoFixture, I think the non-generic `A.Fake` might provide marginal benefit, but if one's already building a framework to work with FakeItEasy, I'm not sure it's a huge advantage.
For end users, I've not yet seen a compelling use case. Once one created the Fake, configuring it would be quite awkward and error-prone.

**Comment by thomaslevesque** [comment]
> You can change your Visual Studio settings so that such members still show up

Yes, but in this case it's the user's choice; he explicitly asked to see all members.

> and IIRC; ReSharper doesn't respect that attribute at all.

It does (at least R# 9.1 does):
![http://i.imgur.com/4yABDcM.png](http://i.imgur.com/4yABDcM.png)

**Comment by adamralph** [comment]
@thomaslevesque in the use case you linked to on SO (http://stackoverflow.com/questions/29902023/can-i-configure-unity-to-automatically-return-a-fake-for-interfaces-that-are-not/29905525#29905525) I don't see you passing a type parameter in the example usage. Am I missing something?

Also, given that what you seem to be trying to do here is just have automatically resolved deps for your fakes, why is Unity a constraint? Is there something you require which, for example, http://www.nuget.org/packages/AutoFixture.AutoFakeItEasy cannot do?

**Comment by blairconrad** [comment]
@adamralph, I think the type comes from

```c#
var type = context.BuildKey.Type;
```

**Comment by adamralph** [comment]
OK, but that is only relevant if Unity is a constraint. Why is it a constraint?

The problem I am seeing here has been solved several times already by various auto faking packages.

**Comment by blairconrad** [comment]
You'll hear no argument from me.

**Comment by thomaslevesque** [comment]
> OK, but that is only relevant if Unity is a constraint. Why is it a constraint?

Well, I thought it was a constraint, but not anymore. I'll probably use Autofixture instead.

> The problem I am seeing here has been solved several times already by various auto faking packages.

Yes, and I suspect most of them solved it using the same ugly reflection hack that I used...

**Comment by adamralph** [comment]
@thomaslevesque OK, thanks - if you need anything else here please feel free to re-open.

**Comment by ploeh** [comment]
> and I suspect most of them solved it using the same ugly reflection hack that I used...

Yes, [pretty much](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFakeItEasy/FakeItEasyMethodQuery.cs#L47-L80) :sweat_smile:

**Comment by ulrichb** [comment]
Hi. I would like to bring in a new use case into the discussion. The use case just popped up in a project. The following snippets show a simplified version of it.


Given a "special setup" type _hierarchy_.
```C#
public interface ISomething
{
  ConcreteNeedsSpecialSetup Property { get; }
}

public abstract class NeedsSpecialSetup
{
  protected NeedsSpecialSetup(string a)
  {
    if (a != "special c'tor value")
          throw new Exception();
  }
}

public class ConcreteNeedsSpecialSetup : NeedsSpecialSetup
{
  public ConcreteNeedsSpecialSetup(string a) : base(a) { }
}
```

... then you can't just rely on FIE's default fake behavior, because the dummy creation in FIE fails:
```C#
var fake = A.Fake<ISomething>();

// Doesn't work because when FIE tries to create a dummy "ConcreteNeedsSpecialSetup" 
// for "Property", the c'tor throws an exception:
A.CallTo(() => fake.Property.ToString()).Returns("faked value");

Assert.That(fake.Property.ToString(), Is.EqualTo("faked value"));
```

=> We write a FIE `IDummyFactory`. The result values of the `IDummyFactory` have to be fakes because we want to influence their virtual members. *Because the `IDummyFactory` should be implemented for the whole `NeedsSpecialSetup` hierarchy we would now need the non-generic overload `A.Fake(Type type, Action<IFakeOptions>)`.* At the moment, the following workaround is necessary.
```C#
public class MyDummyFactory : IDummyFactory
{
  public Priority Priority => Priority.Default;

  public bool CanCreate(Type type) => typeof(NeedsSpecialSetup).IsAssignableFrom(type);

  public object Create(Type type)
  {
    var methodInfo = typeof(MyDummyFactory).GetMethod(nameof(CreateNeedsSpecialSetup), 
      BindingFlags.NonPublic | BindingFlags.Static);
    var boundDomainObject = methodInfo.MakeGenericMethod(type);
    return boundDomainObject.Invoke(null, new object[0]);
  }

  private static T CreateNeedsSpecialSetup<T>() where T : NeedsSpecialSetup
  {
    return A.Fake<T>(o => o.WithArgumentsForConstructor(new[] { "special c'tor value" }));
  }
}

```

**Comment by blairconrad** [comment]
Thanks for chiming in, @ulrichb. 

(First, an aside. I'm excited to see someone using the more powerful `DummyFactory`s. I'm glad they're of use.)

Your example seems to be a case of constructing your own mini-framework in the tests, but I do see how you were inconvenienced. If the non-generic overload were supported, you'd be able to write, I'm thinking something like this?

```c#
public class MyDummyFactory : IDummyFactory
{
    public Priority Priority => Priority.Default;

    public bool CanCreate(Type type) => typeof(NeedsSpecialSetup).IsAssignableFrom(type);

    public object Create(Type type)
    {
       return A.Fake(type, o => o.WithArgumentsForConstructor(new[] { "special c'tor value" }));
    }
}
```

I can see the appeal. I think I need more time to think, but I'm being swayed somewhat.

**Comment by blairconrad** [comment]
Oh, and now that we have a non-generic `IFakeOptions`, I think @thomaslevesque [comment](https://github.com/FakeItEasy/FakeItEasy/issues/482#issuecomment-96871189) is less of a barrier (none?).

**Comment by thomaslevesque** [comment]
> Oh, and now that we have a non-generic `IFakeOptions`, I think @thomaslevesque comment is less of a barrier (none?).

Yes. I had forgotten about it, and was about to suggest that we would need to introduce it ;)

**Comment by thomaslevesque** [comment]
@FakeItEasy/owners, do we want to reopen this? (I do, obviously, since I was the one who opened the issue initially :wink:)

**Comment by blairconrad** [comment]
I'm for reopening.

One oddball person coming our of nowhere and requesting the feature is one matter, but now that we have two such people, it appears that the feature would fill a desperate need. And who knows, maybe someday @ploeh will not mind taking a dependency on a modern FakeItEasy and he can have prettier code in AutoFixture.

**Comment by ulrichb** [comment]
@blairconrad  
> you'd be able to write, I'm thinking something like this?

Yes, exactly. This would be possible with the overload mentioned in my first post.

Obviously, I'm also for reopening.

**Comment by thomaslevesque** [comment]
I just started working on this, it shouldn't take very long.

**Comment by thomaslevesque** [comment]
While I'm at it, should I also add non-generic overloads for the following methods?

- `A.Dummy`
- `A.CollectionOfFake`
- `A.CollectionOfDummy`

**Comment by adamralph** [comment]
Makes sense to me.

BTW - what happens to Intellisense when these new overloads are added? Does it add any clutter which gets in the way of `A.Fake<T>`? Given that at least 99.999% of users are going to keep using `A.Fake<T>` we need to make sure we don't degrade the experience for them.

**Comment by thomaslevesque** [comment]
> BTW - what happens to Intellisense when these new overloads are added? Does it add any clutter which gets in the way of A.Fake<T>? Given that at least 99.999% of users are going to keep using A.Fake<T> we need to make sure we don't degrade the experience for them.

Good point. With VS Intellisense, the non-generic overload appears first, which is not what we want... (with R# Intellisense, the generic overload appears first). Possible solutions:
- hide the new overloads from Intellisense, since they're only useful in very specific scenarios... but on the other hand, people who need them will assume they don't exist if they don't appear in Intellisense. Or we could use `[EditorBrowsable(EditorBrowsableState.Advanced)]`, rather than `Never`.
- Don't expose them directly on `A`; we could do something like `A.NonGeneric.Fake(typeof(Foo))`. But it doesn't read well at all...

Not sure what the best course is here...

**Comment by thomaslevesque** [comment]
> With VS Intellisense, the non-generic overload appears first, which is not what we want

Actually, I'm not sure it's really an issue, because VS Intellisense never inserts the `<>` characters upon completion, even when the generic overload is selected. So whether the user selects the generic or non-generic method, they'll still need to insert the `<>` manually.

**Comment by adamralph** [comment]
IMHO `EditorBrowsable` isn't a solution. We can't rely on it being respected in all editors, with all plugins, on all platforms (a continually growing ecosystem).

I'll grab your branch and take a look...

**Comment by thomaslevesque** [comment]
> I'll grab your branch and take a look...

I haven't pushed anything yet. But you can easily test it by adding the methods manually:

        public static object Fake(Type typeOfFake)
        {
            return FakeCreator.CreateFake(typeOfFake, x => { });
        }

        public static object Fake(Type typeOfFake, Action<IFakeOptions> optionsBuilder)
        {
            return FakeCreator.CreateFake(typeOfFake, optionsBuilder);
        }

**Comment by thomaslevesque** [comment]
I pushed what I've done so far. Still missing specs for `A.Dummy`/`A.CollectionOfDummy`; I don't think we have any specs covering the generic `A.Dummy<T>` yet.

**Comment by blairconrad** [comment]
Looking pretty good, @thomaslevesque. I did add some fairly superficial comments.

**Comment by adamralph** [comment]
@blairconrad indeed!

**Comment by adamralph** [comment]
Reopening, since we have an unresolved concern over Intellisense degradation.

**Comment by thomaslevesque** [comment]
> Reopening, since we have an unresolved concern over Intellisense degradation.

It didn't seem to be much of an issue to me... have you tried it?

**Comment by adamralph** [comment]
Yes I tried it, that's why I raised a concern :wink:

**Comment by thomaslevesque** [comment]
So... what exactly bothered you? And what should we do about it?

**Comment by thomaslevesque** [comment]
(forget I asked, I just noticed the discussion in the chat)

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.1.0).
